src/tqec/sketchup/zx_graph.py

```diff
@@ -0,0 +1,334 @@
+"""ZX graph representation of a 3D spacetime defect diagram."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from enum import Enum, auto
+from dataclasses import dataclass, astuple
+
+import networkx as nx
+import numpy as np
+
+from tqec.exceptions import TQECException
+
+if TYPE_CHECKING:
+    from tqec.sketchup.block_graph import BlockGraph
+
+
+class NodeType(Enum):
+    """Valid node types in a ZX graph."""
+
+    X = "x"  # X-type node
+    Z = "z"  # Z-type node
+    V = "v"  # Virtual node that represents an open port
+
+
+@dataclass(frozen=True, order=True)
+class Position3D:
+    """A 3D integer position."""
+
+    x: int
+    y: int
+    z: int
+
+    def shift_by(self, dx: int = 0, dy: int = 0, dz: int = 0) -> Position3D:
+        """Shift the position by the given offset."""
+        return Position3D(self.x + dx, self.y + dy, self.z + dz)
+
+    def is_nearby(self, other: Position3D) -> bool:
+        """Check if the other position is near to this position, i.e. Manhattan distance is 1."""
+        return (
+            abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z) == 1
+        )
```

Comment:
Small nitpick: I think `is_neighbour` would be a better naming, or even `is_direct_neighbour`. I am not a native english speaker, so the probability that I am wrong is high, but to me, "nearby" is looser then "neighbour" and does not really imply that "nearby" nodes are neighbours, they might be 2 edges appart and that would still be a valid "nearby" node.

Reply:
Disregard that, I will do a proper review when you think the PR is ready :)